File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_3_lsp/after/Coach.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_3_lsp.after;

/**
 * The Confident Consumer.
 * * ARCHITECTURE NOTE: Because we strictly adhered to LSP, the Coach class is 
 * incredibly simple and clean. 
 * * Notice what is MISSING here: There are no "if (fieldPlayer instanceof Goalie)" checks. 
 * There are no try/catch blocks expecting an UnsupportedOperationException. The Coach 
 * trusts the abstraction 100%. If an object is passed in as a Player, the Coach 
 * knows for a fact it can play the field.
 */
public class Coach {
    
    /**
     * Directs the player to take their field position with absolute confidence.
     * * @param fieldPlayer A guaranteed field-capable player.
     */
    public void directFieldPlay(Player fieldPlayer) {
        System.out.println("  [Coach] Alright player, execute your field assignment!");
        
        // This will now ALWAYS succeed. No exceptions, no unexpected behavior.
        fieldPlayer.playFieldPosition();
    }
}

File: Chapter 03/C#/Section_3_3_3_LSP/After/Coach.cs
using System;

namespace Chapter03.LSP.After;

/// <summary>
/// The Confident Consumer.
/// 
/// ARCHITECTURE NOTE: Because we strictly adhered to LSP, the Coach class is 
/// incredibly simple and clean. 
/// 
/// Notice what is MISSING here: There are no "if (fieldPlayer is Goalie)" checks. 
/// There are no try/catch blocks expecting a NotImplementedException. The Coach 
/// trusts the abstraction 100%. If an object is passed in as a Player, the Coach 
/// knows for a fact it can play the field.
/// </summary>
public class Coach
{
    /// <summary>
    /// Directs the player to take their field position with absolute confidence.
    /// </summary>
    /// <param name="fieldPlayer">A guaranteed field-capable player.</param>
    public void DirectFieldPlay(Player fieldPlayer)
    {
        Console.WriteLine("  [Coach] Alright player, execute your field assignment!");
        
        // This will now ALWAYS succeed. No exceptions, no unexpected behavior.
        fieldPlayer.PlayFieldPosition();
    }
}